### PR TITLE
Introduce the ViewComponent pattern

### PR DESCRIPTION
**Pull request by sogamoso**
Co-authored by: Dyego Costa <dyego@dyegocosta.com>

### CODE REVIEW

**Review by RenatoCoelho-BiggerPockets — APPROVED** [review]
Great job on this @sogamoso and @DyegoCosta

**Review by bhanubhakta — APPROVED** [review]
Great job ❤️

**Review by baerbradford — APPROVED** [review]
Looks great, thanks for writing this up!

Approved with the one caveat that we need to update the lookbook URL once we have that available.

**Comment by sogamoso** [comment]
> Approved with the one caveat that we need to update the lookbook URL once we have that available.

@baerbradford, updated!

File: _patterns/build-views-with-view-components.md [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+* Partials
+* ERB files with business logic
+* Helpers
+* Decorators
+
+### Flexible components
+
+Make sure to generalize your use case so that the component can be used in any
+part of the application.
+
```
Comment (bockets) [inline _patterns/build-views-with-view-components.md:18]: I do think it's OK to start with a specific component and then extract the reusable bits. Then again, maybe if it's too-specific, it shouldn't be a component. I just want to encourage people to stick to the middle ground: use a view component if you think others could benefit from reusing it. Otherwise, use a partial or a namespaced component.

File: _patterns/build-views-with-view-components.md [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+In order to make components more reusable the library provides us with concepts like
+the `content` helper and [slots](https://viewcomponent.org/guide/slots.html).
+
+Always prefer extending component or composing new ones with existing ones over
```
Comment (bockets) [inline _patterns/build-views-with-view-components.md:22]:
```suggestion
Prefer extending component or composing new ones with existing ones over 
```

File: _patterns/build-views-with-view-components.md [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+In order to make components more reusable the library provides us with concepts like
+the `content` helper and [slots](https://viewcomponent.org/guide/slots.html).
+
+Always prefer extending component or composing new ones with existing ones over 
+creating additional components. Feature specific components are welcomed, but they 
+should be composed by other reusable components. There are more details on this topic 
+[here](https://viewcomponent.org/viewcomponents-at-github.html#the-two-types-of-viewcomponents-we-write).
+
+Here's an example that help illustrate how components can be flexible:
```
Comment (bockets) [inline _patterns/build-views-with-view-components.md:27]:
```suggestion
Here's an example that illustrates how components can be flexible:
```

File: _patterns/build-views-with-view-components.md [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+<%= render ButtonWithIcon.new(text: "Button name", icon: "assets/icons/star.svg") %>
+```
+
+#### Good
+```rb
+<%= render Button.new(text: "Big button", size: :large) %>
+<%= render Button.new(text: "Button name") do |component| %>
+  <% component.with_leading_icon(icon: "assets/icons/star.svg") %>
+<% end %>
+```
```
Comment (bockets) [inline _patterns/build-views-with-view-components.md:41]: Huh, I would have thought that the icon-ness of a Button would be best expressed with a simple named argument, but I guess this compositional style allows you to place the icon anywhere.

I think that rubyists shy away from too many named args too much.

Where does the `with_leading_icon` get declared? In this example, it looks generic: _any_ component could take that as an argument, but does it really work for any component? My preference would be:

```rb
Button.new(text: "Button name", with_leading_icon: "assets/icon/star.svg")

File: _patterns/build-views-with-view-components.md [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+
+### Styled components
+
+Avoid using writting CSS when building view components. We're moving away
+from using custom CSS files and using [Tailwind](https://tailwindcss.com) 
+instead. If you're having a hard time styling a component using Tailwind reach 
+out to someone from the design team to discuss making adjustments to the 
+component.
+
+Here's an example that help illustrate how using Tailwind looks as opposed
```
Comment (bockets) [inline _patterns/build-views-with-view-components.md:88]:
```suggestion
Here's an example that illustrates how using Tailwind looks as opposed
```

File: _patterns/build-views-with-view-components.md [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+  <span class="button-content">
+    <span class="button-label">Button text</span>
+  </span>
+</button>
+```
+
+#### Good
+```rb
+# app/components/button.html.erb
+<button type="button" class="inline-flex items-center rounded-md border border-transparent bg-indigo-600 px-4 py-2 text-sm font-medium text-white shadow-sm">
```
Comment (bockets) [inline _patterns/build-views-with-view-components.md:111]: Missing dark-mode :trollface:

File: _patterns/build-views-with-view-components.md [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+  Button text
+</button>
+```
+
+### Shared components
+
+Always check [our custom component library](https://www.biggerpockets.com/lookbook) 
+before creating a new one. The goal of our component library is not to grow 
+indefinitely but instead to provide the building blocks to implement the view layer 
+of all user facing features.
```
Comment (bockets) [inline _patterns/build-views-with-view-components.md:121]:
```suggestion
before creating a new one. The goal of our component library is to succinctly provide
the minimum building blocks necessary for implementing the view layer of all user facing features.
```

**Review by bockets — COMMENTED** [review]
Excellent pattern

File: _patterns/build-views-with-view-components.md [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+creating additional components. Feature specific components are welcomed, but they 
+should be composed by other reusable components. There are more details on this topic 
+[here](https://viewcomponent.org/viewcomponents-at-github.html#the-two-types-of-viewcomponents-we-write).
+
+Here's an example that help illustrate how components can be flexible:
+
+#### Bad
+```rb
+<%= render BigButton.new(text: "Big button") %>
+<%= render ButtonWithIcon.new(text: "Button name", icon: "assets/icons/star.svg") %>
```
Comment (RenatoCoelho-BiggerPockets) [inline _patterns/build-views-with-view-components.md:32]:
```suggestion
<%= render ButtonWithIcon.new(text: "Button with icon", icon: "assets/icons/star.svg") %>
```

File: _patterns/build-views-with-view-components.md [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+```
+
+#### Good
+```rb
+<%= render Button.new(text: "Big button", size: :large) %>
+<%= render Button.new(text: "Button name") do |component| %>
```
Comment (RenatoCoelho-BiggerPockets) [inline _patterns/build-views-with-view-components.md:38]:
```suggestion
<%= render Button.new(text: "Button with icon") do |component| %>
```

File: _patterns/build-views-with-view-components.md (hunk 4 above)
Comment (RenatoCoelho-BiggerPockets) [reply _patterns/build-views-with-view-components.md:41]: Dyego created an Icon component that receives an icon_name as an input and used it in the SelectableCards component by passing it the icon_name [here](https://github.com/BiggerPockets/biggerpockets/blob/96a654dc02a349db634fa7d080d8d99c8e613437/app/components/thumbnail_box_component.rb#L7), so wdyt about changing this to match what is implemented?

Comment (bockets) [reply _patterns/build-views-with-view-components.md:41]: I like Dyego's solution. I think positioning an icon in a button can be done two ways: composition or an argument. I like both, and maybe we support both, but there should be a simple happy path for prepending/appending an icon to a button since it will be done again and again and again.

File: _patterns/build-views-with-view-components.md [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+```rb
+# app/components/button.html.erb
+<button type="button" class="button">
+  <span class="button-content">
+    <span class="button-label">Button text</span>
+  </span>
+</button>
+```
+
+#### Good
```
Comment (johngallagher) [inline _patterns/build-views-with-view-components.md:108]: Small tweak - can we consistently have blank lines after headings in Markdown? Makes no difference to the preview but makes it much easier on the eye to edit...

```suggestion
#### Good

```

File: _patterns/build-views-with-view-components.md [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+before creating a new one. The goal of our component library is not to grow 
+indefinitely but instead to provide the building blocks to implement the view layer 
+of all user facing features.
+
+To make sure new components are included in the component library
+a [component preview](https://viewcomponent.org/guide/previews.html) is required. 
+Previews should include [annotations](https://lookbook.build/guide/previews/annotating/) 
+and parameter descriptions. It should also ideally present the default use of 
+the component along with common variations. Component previews automatically get 
+included in the Lookbook.
```
Comment (johngallagher) [inline _patterns/build-views-with-view-components.md:128]: To make this more scannable, convert to bullets:

```suggestion
* To make sure new components are included in the component library a [component preview](https://viewcomponent.org/guide/previews.html) is required
* Previews should include [annotations](https://lookbook.build/guide/previews/annotating/) 
and parameter descriptions
* Annotations should present the default use of the component along with common variations
* Component previews automatically get included in the Lookbook
```

**Review by johngallagher — APPROVED** [review]
Some small tweaks, but overall it looks great! Thank you!

File: _patterns/build-views-with-view-components.md [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+```
+
+#### Good
+```rb
+# app/components/button.html.erb
+<button type="button" class="inline-flex items-center rounded-md border border-transparent bg-indigo-600 px-4 py-2 text-sm font-medium text-white shadow-sm">
+  Button text
+</button>
+```
+
```
Comment (DyegoCosta) [inline _patterns/build-views-with-view-components.md:115]: I couldn't use a code block inside the suggestion block

```suggestion

### Dynamic styling with Tailwind

Some components may require different styling based on their state or arguments passed to them, and this can be a challenge to handle when working Tailwind because it [requires CSS classes to be spelled out](https://github.com/rails/tailwindcss-rails#class-names-must-be-spelled-out).

In order to make this work we need to define and spell out all CSS classes a component can use by using methods on the component ruby class.

#### Bad

This will not work:

```ruby
class Button
  def initialize(scheme:)
    @scheme = scheme
  end
  
  def scheme_color
    scheme == :primary ? "blue" : "white"
  end
end
ruby```


```erb
`<input class="bg-<%= scheme_color %>" />`
erb```

#### Good

```ruby
class Button
  def initialize(scheme:)
    @scheme = scheme
  end
  
  def scheme_color_class
    scheme == :primary ? "bg-blue" : "bg-white"
  end
end
ruby```


```erb
`<input class="<%= scheme_color_class %>" />`
erb```

```

File: _patterns/build-views-with-view-components.md [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+- View component tests are much faster than system tests.
+- These tests allow us to easily check for edge cases.
+
+Here's an example that help illustrate how simple a component test can be:
+
+```rb
+# app/components/table_component.rb
+class TableComponent < ApplicationComponent
+  def initialize(collection:)
+    @collection = collection
```
Comment (bockets) [inline _patterns/build-views-with-view-components.md:63]: Can we use a different name than `collection` for the example? That name implies slots, but the example doesn't use slots.

File: _patterns/build-views-with-view-components.md (hunk 1 above)
Comment (sogamoso) [reply _patterns/build-views-with-view-components.md:18]: >  do think it's OK to start with a specific component and then extract the reusable bits. Then again, maybe if it's too-specific, it shouldn't be a component. I just want to encourage people to stick to the middle ground: use a view component if you think others could benefit from reusing it.

@bockets, the goal is to encourage us to think about how we can make our work more reusable instead of blindly creating a feature specific and missing the opportunity to save time for our future selves.

**Review by BPMattBauer — APPROVED** [review]

**Review by johngallagher — APPROVED** [review]

**Review by DyegoCosta — APPROVED** [review]

File: _patterns/build-views-with-view-components.md (hunk 4 above)
Comment (sogamoso) [reply _patterns/build-views-with-view-components.md]: Done!

File: _patterns/build-views-with-view-components.md (hunk 11 above)
Comment (sogamoso) [reply _patterns/build-views-with-view-components.md]: Done!
